webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/ncf/TechniqueReader.scala

```diff
@@ -53,11 +53,21 @@ class TechniqueReader(
     }
   }
 
+  private[this] def doesMethodsMetadataFileNeedsUpdate : Boolean = {
+    val methodsFileModifiedTime = methodsFile.lastModifiedTime()
+    def isAMethodNewerThanCache(file : File) : Boolean = {
+      file.isRegularFile && file.extension(false).map(_ == "cf").getOrElse(false)  && file.lastModifiedTime.isAfter(methodsFileModifiedTime)
+    }
+    val baseDir = root / "usr" / "share" / "ncf" / "tree" / "30_generic_methods"
+    val userDir = configuration_repository / "ncf" / "30_generic_methods"
+    baseDir.collectChildren(isAMethodNewerThanCache).isEmpty || userDir.collectChildren(isAMethodNewerThanCache).isEmpty
+  }
```

Comment:
I think it's ok but it's hard to test as it is. What would make it easier to test is to have it split in two: 
- one part with the logic and parameters for all states, 
- one methods that set state for user convenience, ie:

```
 // this one contains logic but stateless and so easy to test - actually, we could make it pure
 // with an `IORestul.effect` around last statement
  private[this] def checkNeedsUpdate(lastTime: TimeType, dirs: List[File]) : Boolean = {
    def isAMethodNewerThanCache(file : File) : Boolean = {
      file.isRegularFile && file.extension(false).map(_ == "cf").getOrElse(false)  && lastTime.isAfter(methodsFileModifiedTime)
    }
    dirs.exist(dir => dir.collectChildren(isAMethodNewerThanCache).isEmpty)
  }
  // this one is convenient and meaningful to use in the code (but untestable)
  private[this] def doesMethodsMetadataFileNeedsUpdate : Boolean = {
    val methodsFileModifiedTime = methodsFile.lastModifiedTime()
    val baseDir = root / "usr" / "share" / "ncf" / "tree" / "30_generic_methods"
    val userDir = configuration_repository / "ncf" / "30_generic_methods"
   checkNeedsUpdate(methodsFileModifiedTime, baseDir :: userDir :: Nil)
  }
```